colexec: diskqueue nil pointer dereference

DiskQueue runs into nil pointer dereference error when running the sqlite logic test in [`select4.test`][1]

Using the tables created in `select4.test` , following query causes nil pointer dereference error:

```
SELECT e8, d7+959, a5*693, e2, e6*447, b1+778
  FROM t1, t6, t7, t5, t8, t2
 WHERE c5 in (489,250,927,855,697,130,856)
   AND (180=e8 OR 333=e8 OR e8=38)
   AND e7=462
   AND b2=b1
```

Error stack trace: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5be6121]

goroutine 3154 [running]:
github.com/cockroachdb/cockroach/pkg/col/colserde.(*countingWriter).Write(...)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/col/colserde/file.go:329
github.com/cockroachdb/cockroach/pkg/col/colserde.(*FileSerializer).Finish(0xc0054f3980, 0x0, 0x0)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/col/colserde/file.go:139 +0x131
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).writeFooterAndFlush(0xc004052480, 0x7f0dd20, 0xc003889400, 0xc00226dc10, 0x400855f)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:481 +0x36
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close(0xc004052480, 0x7f0dd20, 0xc003889400, 0xc00226dcc8, 0x404252f)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:399 +0x36c
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*spillingQueue).close(0xc00328cc00, 0x7f0dd20, 0xc003889400, 0x4, 0xc003a29e60)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:271 +0x65
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).closeLocked(0xc0021f99e0, 0x7f0dd20, 0xc003889400)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:225 +0x4b
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).cancel(0xc0021f99e0, 0x7f0dd20, 0xc003889400)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:238 +0x58
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run.func1(0x7e9e3e0, 0xc005892b70)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:457 +0xa3
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run(0xc0021f9b00, 0x7f0dd20, 0xc003889400)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:497 +0x1b6
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1(0x7f0dd20, 0xc003889400, 0xc001ce1450)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:676 +0xa5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc003a6e1a0, 0x7f0dd20, 0xc003889400, 0xc001ce1450, 0xc004052f70)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1171 +0x44
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1170 +0x6f


```

[1]: https://github.com/cockroachdb/sqllogictest/blob/master/test/select4.test

The stacktrace above is actually misleading. 

We catch the runtime error inside hash router [here][1] and attempt to cancel all `routerOutputOp` the router is attached to. However at this point the nil pointer dereference has already happened. The stack trace above comes from the attempt by hash router to cancel operations. 

More accurate stack trace: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x5be4f6a]

goroutine 3128 [running]:
github.com/cockroachdb/cockroach/pkg/col/colserde.(*FileSerializer).AppendBatch(0xc0032b53e0, 0x7f75f40, 0xc0029d51c0, 0x5bf900b, 0xc0029d51c0)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/col/colserde/file.go:108 +0x2a
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Enqueue(0xc0034fa000, 0x7f0cce0, 0xc002532580, 0x7f75f40, 0xc0029d51c0, 0xc0029d53c0, 0xc0028e0870)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:541 +0x29a
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*spillingQueue).enqueue(0xc000e58f00, 0x7f0cce0, 0xc002532580, 0x7f75f40, 0xc0029d51c0, 0x7f75f40, 0xc0029d51c0)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:144 +0x166
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).addBatch(0xc0035ff5f0, 0x7f0cce0, 0xc002532580, 0x7f75f40, 0xc003203280, 0xc001259bc8, 0x1, 0x1, 0xc0020fc400)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:313 +0x3db
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).processNextBatch(0xc0035ff9e0, 0x7f0cce0, 0xc002532580, 0x73a4500)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:527 +0x2ca
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run.func2(...)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:462
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run(0xc0035ff9e0, 0x7f0cce0, 0xc002532580)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:496 +0x111
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1(0x7f0cce0, 0xc002532580, 0xc0025b83f0)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:676 +0xa5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc0033605e0, 0x7f0cce0, 0xc002532580, 0xc0025b83f0, 0xc0007128b0)
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1171 +0x44
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/Users/azhng/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1170 +0x6f


```

[1]: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/colexec/routers.go#L496

This problem appears on all of the configurations, and it seems like an issue with the test harness:
```
    --- PASS: TestLogic/local (0.00s)
    --- FAIL: TestLogic/local-vec-off (0.00s)
        --- FAIL: TestLogic/local-vec-off/temp (2.35s)
            logic.go:1858: 
                 
                testdata/logic_test/temp:3095: expected 18432 results, but found 25344
    --- PASS: TestLogic/local-v1.1@v1.0-noupgrade (0.00s)
    --- FAIL: TestLogic/local-vec (0.00s)
        --- FAIL: TestLogic/local-vec/temp (0.73s)
            logic.go:1276: randomize batchSize to 1024
                
            logic.go:1858: 
                 
                testdata/logic_test/temp:3095: expected 18432 results, but found 25344
    --- PASS: TestLogic/fakedist (0.00s)
    --- PASS: TestLogic/local-mixed-19.2-20.1 (0.00s)
    --- FAIL: TestLogic/fakedist-vec-off (0.00s)
        --- FAIL: TestLogic/fakedist-vec-off/temp (2.38s)
            logic.go:1858: 
                 
                testdata/logic_test/temp:3095: expected 18432 results, but found 25344
    --- FAIL: TestLogic/fakedist-vec (0.00s)
        --- FAIL: TestLogic/fakedist-vec/temp (1.29s)
            logic.go:1276: randomize batchSize to 1024
                
            logic.go:1858: 
                 
                testdata/logic_test/temp:3095: expected 18432 results, but found 25344
    --- FAIL: TestLogic/fakedist-vec-disk (0.00s)
        --- FAIL: TestLogic/fakedist-vec-disk/temp (1.88s)
            logic.go:1276: randomize batchSize to 1024
                
            logic.go:1858: 
                 
                testdata/logic_test/temp:3095: expected 18432 results, but found 25344
```
I'll try to find the discussion that we had about this column number mismatch previously.

--- a/pkg/sql/colcontainer/diskqueue.go
+++ b/pkg/sql/colcontainer/diskqueue.go
@@ -20,7 +20,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/col/coldata"
 	"github.com/cockroachdb/cockroach/pkg/col/colserde"
 	"github.com/cockroachdb/cockroach/pkg/col/coltypes"
-	"github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror"
 	"github.com/cockroachdb/cockroach/pkg/storage/fs"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
@@ -425,6 +424,9 @@ func (d *diskQueue) Close(ctx context.Context) error {
 	for _, file := range d.files[leftOverFileIdx : d.writeFileIdx+1] {
 		totalSize += int64(file.totalSize)
 	}
+	if totalSize > d.diskAcc.Used() {
+		totalSize = d.diskAcc.Used()
+	}
 	d.diskAcc.Shrink(ctx, totalSize)
 	return nil
 }
@@ -489,7 +491,7 @@ func (d *diskQueue) writeFooterAndFlush(ctx context.Context) error {
 	d.numBufferedBatches = 0
 	d.files[d.writeFileIdx].totalSize += written
 	if err := d.diskAcc.Grow(ctx, int64(written)); err != nil {
-		execerror.VectorizedInternalPanic(err)
+		return err
 	}
 	// Append offset for the readers.
 	d.files[d.writeFileIdx].offsets = append(d.files[d.writeFileIdx].offsets, d.files[d.writeFileIdx].totalSize)
@@ -578,7 +580,11 @@ func (d *diskQueue) maybeInitDeserializer(ctx context.Context) (bool, error) {
 				if err := d.cfg.FS.DeleteFile(d.files[d.readFileIdx].name); err != nil {
 					return false, err
 				}
-				d.diskAcc.Shrink(ctx, int64(d.files[d.readFileIdx].totalSize))
+				fileSize := int64(d.files[d.readFileIdx].totalSize)
+				if fileSize > d.diskAcc.Used() {
+					fileSize = d.diskAcc.Used()
+				}
+				d.diskAcc.Shrink(ctx, fileSize)
 			}
 			d.readFile = nil
 			// Read next file.

--- a/pkg/sql/colexec/routers.go
+++ b/pkg/sql/colexec/routers.go
@@ -38,6 +38,8 @@ type routerOutput interface {
 	addBatch(context.Context, coldata.Batch, []int) bool
 	// cancel tells the output to stop producing batches.
 	cancel(ctx context.Context)
+	// drain drains the output of any metadata.
+	drain() []execinfrapb.ProducerMetadata
 }
 
 // getDefaultRouterOutputBlockedThreshold returns the number of unread values
@@ -48,6 +50,19 @@ type routerOutput interface {
 func getDefaultRouterOutputBlockedThreshold() int {
 	return coldata.BatchSize() * 2
 }
+
+type routerOutputOpState int
+
+const (
+	// routerOutputOpRunning is the state in which routerOutputOp operates
+	// normally. The router output transitions into the draining state when
+	// either it is finished (when a zero-length batch was added or when it was
+	// canceled) or it encounters an error.
+	routerOutputOpRunning routerOutputOpState = iota
+	// routerOutputOpDraining is the state in which routerOutputOp always
+	// returns zero-length batches on calls to Next.
+	routerOutputOpDraining
+)
 
 type routerOutputOp struct {
 	// input is a reference to our router.
@@ -61,6 +76,7 @@ type routerOutputOp struct {
 
 	mu struct {
 		syncutil.Mutex
+		state routerOutputOpState
 		// unlimitedAllocator tracks the memory usage of this router output,
 		// providing a signal for when it should spill to disk.
 		// The memory lifecycle is as follows:
@@ -92,7 +108,6 @@ type routerOutputOp struct {
 		// are released.
 		unlimitedAllocator *Allocator
 		cond               *sync.Cond
-		done               bool
 		// pendingBatch is a partially-filled batch with data added through
 		// addBatch. Once this batch reaches capacity, it is flushed to data. The
 		// main use of pendingBatch is coalescing various fragmented batches into
@@ -102,6 +117,17 @@ type routerOutputOp struct {
 		data      *spillingQueue
 		numUnread int
 		blocked   bool
+
+		drainState struct {
+			// err stores any error (if such occurs) when dequeueing from data
+			// (one possible error could be hitting disk usage limit). Once
+			// such error occurs, the output transitions into draining state.
+			// The error, however, will not be propagated right away - in order to
+			// prevent double reporting of the error by the hash router, it will be
+			// returned either on the next call to addBatch (if such occurs) or
+			// during draining of the router output.
+			err error
+		}
 	}
 
 	testingKnobs struct {
@@ -182,14 +208,11 @@ func (o *routerOutputOp) Init() {}
 func (o *routerOutputOp) Next(ctx context.Context) coldata.Batch {
 	o.mu.Lock()
 	defer o.mu.Unlock()
-	if o.mu.done {
-		return coldata.ZeroBatch
-	}
-	for o.mu.pendingBatch == nil && o.mu.data.empty() && !o.mu.done {
+	for o.mu.state == routerOutputOpRunning && o.mu.pendingBatch == nil && o.mu.data.empty() {
 		// Wait until there is data to read or the output is canceled.
 		o.mu.cond.Wait()
 	}
-	if o.mu.done {
+	if o.mu.state == routerOutputOpDraining {
 		return coldata.ZeroBatch
 	}
 	var b coldata.Batch
@@ -201,10 +224,10 @@ func (o *routerOutputOp) Next(ctx context.Context) coldata.Batch {
 		o.mu.unlimitedAllocator.ReleaseBatch(b)
 		o.mu.pendingBatch = nil
 	} else {
-		var err error
-		b, err = o.mu.data.dequeue(ctx)
-		if err != nil {
-			execerror.VectorizedInternalPanic(err)
+		b, o.mu.drainState.err = o.mu.data.dequeue(ctx)
+		if o.mu.drainState.err != nil {
+			o.mu.state = routerOutputOpDraining
+			return coldata.ZeroBatch
 		}
 	}
 	o.mu.numUnread -= b.Length()
@@ -221,7 +244,7 @@ func (o *routerOutputOp) Next(ctx context.Context) coldata.Batch {
 }
 
 func (o *routerOutputOp) closeLocked(ctx context.Context) {
-	o.mu.done = true
+	o.mu.state = routerOutputOpDraining
 	if err := o.mu.data.close(ctx); err != nil {
 		// This log message is Info instead of Warning because the flow will also
 		// attempt to clean up the parent directory, so this failure might not have
@@ -262,6 +285,16 @@ func (o *routerOutputOp) addBatch(ctx context.Context, batch coldata.Batch, sele
 	}
 	o.mu.Lock()
 	defer o.mu.Unlock()
+	if o.mu.state == routerOutputOpDraining {
+		// We have encountered an error in Next() but have not yet propagated
+		// it, so we do so now - the error will get to the hash router which
+		// will cancel all of its outputs.
+		err := o.mu.drainState.err
+		// We set the error to nil so that it is not propagated again, during
+		// drain() call.
+		o.mu.drainState.err = nil
+		execerror.VectorizedInternalPanic(err)
+	}
 	if batch.Length() == 0 {
 		if o.mu.pendingBatch != nil {
 			if err := o.mu.data.enqueue(ctx, o.mu.pendingBatch); err != nil {
@@ -339,11 +372,23 @@ func (o *routerOutputOp) maybeUnblockLocked() {
 	}
 }
 
+func (o *routerOutputOp) drain() []execinfrapb.ProducerMetadata {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+	if o.mu.drainState.err != nil {
+		err := o.mu.drainState.err
+		o.mu.drainState.err = nil
+		return []execinfrapb.ProducerMetadata{{Err: err}}
+	}
+	return nil
+}
+
 // reset resets the routerOutputOp for a benchmark run.
 func (o *routerOutputOp) reset(ctx context.Context) {
 	o.mu.Lock()
-	o.mu.done = false
+	o.mu.state = routerOutputOpRunning
 	o.mu.data.reset(ctx)
+	o.mu.drainState.err = nil
 	o.mu.numUnread = 0
 	o.mu.blocked = false
 	o.mu.Unlock()
@@ -572,6 +617,9 @@ func (r *HashRouter) reset(ctx context.Context) {
 func (r *HashRouter) DrainMeta(ctx context.Context) []execinfrapb.ProducerMetadata {
 	r.mu.Lock()
 	defer r.mu.Unlock()
+	for _, o := range r.outputs {
+		r.mu.bufferedMeta = append(r.mu.bufferedMeta, o.drain()...)
+	}
 	meta := r.mu.bufferedMeta
 	r.mu.bufferedMeta = r.mu.bufferedMeta[:0]
 	return meta

--- a/pkg/sql/colexec/routers.go
+++ b/pkg/sql/colexec/routers.go
@@ -446,62 +446,82 @@ func newHashRouterWithOutputs(
 // output calculated by hashing columns. Cancel the given context to terminate
 // early.
 func (r *HashRouter) Run(ctx context.Context) {
-	r.input.Init()
-	cancelOutputs := func(err error) {
-		if err != nil {
-			r.mu.Lock()
-			r.mu.bufferedMeta = append(r.mu.bufferedMeta, execinfrapb.ProducerMetadata{Err: err})
-			r.mu.Unlock()
-		}
-		for _, o := range r.outputs {
-			o.cancel(ctx)
-		}
+	bufferErr := func(err error) {
+		r.mu.Lock()
+		r.mu.bufferedMeta = append(r.mu.bufferedMeta, execinfrapb.ProducerMetadata{Err: err})
+		r.mu.Unlock()
 	}
-	var done bool
-	processNextBatch := func() {
-		done = r.processNextBatch(ctx)
-	}
-	for {
-		// Check for cancellation.
-		select {
-		case <-ctx.Done():
-			cancelOutputs(ctx.Err())
-			return
-		default:
-		}
-
-		// Read all the routerOutput state changes that have happened since the
-		// last iteration.
-		for moreToRead := true; moreToRead; {
-			select {
-			case <-r.unblockedEventsChan:
-				r.numBlockedOutputs--
-			default:
-				// No more routerOutput state changes to read without blocking.
-				moreToRead = false
+	// Since HashRouter runs in a separate goroutine, we want to be safe and
+	// make sure that we catch errors in all code paths, so we wrap the whole
+	// method with a catcher. Note that we also have "internal" catchers as
+	// well for more fine-grained control of error propagation.
+	if err := execerror.CatchVectorizedRuntimeError(func() {
+		r.input.Init()
+		// cancelOutputs buffers non-nil error as metadata, cancels all of the
+		// outputs additionally buffering any error if such occurs during the
+		// outputs' cancellation as metadata as well. Note that it attempts to
+		// cancel every output regardless of whether "previous" output's
+		// cancellation succeeds.
+		cancelOutputs := func(err error) {
+			if err != nil {
+				bufferErr(err)
+			}
+			for _, o := range r.outputs {
+				if err := execerror.CatchVectorizedRuntimeError(func() {
+					o.cancel(ctx)
+				}); err != nil {
+					bufferErr(err)
+				}
 			}
 		}
-
-		if r.numBlockedOutputs == len(r.outputs) {
-			// All outputs are blocked, wait until at least one output is unblocked.
+		var done bool
+		processNextBatch := func() {
+			done = r.processNextBatch(ctx)
+		}
+		for {
+			// Check for cancellation.
 			select {
-			case <-r.unblockedEventsChan:
-				r.numBlockedOutputs--
 			case <-ctx.Done():
 				cancelOutputs(ctx.Err())
 				return
+			default:
+			}
+
+			// Read all the routerOutput state changes that have happened since the
+			// last iteration.
+			for moreToRead := true; moreToRead; {
+				select {
+				case <-r.unblockedEventsChan:
+					r.numBlockedOutputs--
+				default:
+					// No more routerOutput state changes to read without blocking.
+					moreToRead = false
+				}
+			}
+
+			if r.numBlockedOutputs == len(r.outputs) {
+				// All outputs are blocked, wait until at least one output is unblocked.
+				select {
+				case <-r.unblockedEventsChan:
+					r.numBlockedOutputs--
+				case <-ctx.Done():
+					cancelOutputs(ctx.Err())
+					return
+				}
+			}
+
+			if err := execerror.CatchVectorizedRuntimeError(processNextBatch); err != nil {
+				cancelOutputs(err)
+				return
+			}
+			if done {
+				// The input was done and we have notified the routerOutputs that there
+				// is no more data.
+				return
 			}
 		}
-
-		if err := execerror.CatchVectorizedRuntimeError(processNextBatch); err != nil {
-			cancelOutputs(err)
-			return
-		}
-		if done {
-			// The input was done and we have notified the routerOutputs that there
-			// is no more data.
-			return
-		}
+	}); err != nil {
+		bufferErr(err)
 	}
 }
 

--- a/pkg/sql/colexec/routers_test.go
+++ b/pkg/sql/colexec/routers_test.go
@@ -529,6 +529,10 @@ func (o callbackRouterOutput) cancel(context.Context) {
 	}
 }
 
+func (o callbackRouterOutput) drain() []execinfrapb.ProducerMetadata {
+	return nil
+}
+
 func TestHashRouterComputesDestination(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	ctx := context.Background()

